src/cbc_sdk/platform/policy_ruleconfigs.py

```diff
@@ -272,3 +299,316 @@ def set_assignment_mode(self, mode):
         if mode not in ("REPORT", "BLOCK"):
             raise ApiError(f"invalid assignment mode: {mode}")
         self.set_parameter("WindowsAssignmentMode", mode)
+
+
+class HostBasedFirewallRuleConfig(PolicyRuleConfig):
+    """Represents a host-based firewall rule configuration in the policy."""
+    swagger_meta_file = "platform/models/policy_ruleconfig.yaml"
+
+    def __init__(self, cb, parent, model_unique_id=None, initial_data=None, force_init=False, full_doc=False):
+        """
+        Initialize the CorePreventionRuleConfig object.
+
+        Args:
+            cb (BaseAPI): Reference to API object used to communicate with the server.
+            parent (Policy): The "parent" policy of this rule configuration.
+            model_unique_id (str): ID of the rule configuration.
+            initial_data (dict): Initial data used to populate the rule configuration.
+            force_init (bool): If True, forces the object to be refreshed after constructing.  Default False.
+            full_doc (bool): If True, object is considered "fully" initialized. Default False.
+        """
+        super(HostBasedFirewallRuleConfig, self).__init__(cb, parent, model_unique_id, initial_data, force_init,
+                                                          full_doc)
+        self._rule_groups = []
+        self._rule_groups_valid = False
+
+    class FirewallRuleGroup(MutableBaseModel):
+        """Represents a group of related firewall rules."""
+        swagger_meta_file = "platform/models/firewall_rule_group.yaml"
+
+        def __init__(self, cb, initial_data):
+            """
+            Initialize the FirewallRuleGroup object.
+
+            Args:
+                cb (BaseAPI): Reference to API object used to communicate with the server.
+                initial_data (dict): Initial data used to populate the firewall rule group.
+            """
+            super(HostBasedFirewallRuleConfig.FirewallRuleGroup, self).__init__(cb, None, initial_data, False, True)
+            self._parent = None
+            self._rules = [HostBasedFirewallRuleConfig.FirewallRule(cb, d)
+                           for d in initial_data.get("rules", [])]
+
+        def _init_parent(self, parent):
+            """
+            Initialize the parent of this group and its rules.
+
+            Args:
+                parent (HostBasedFirewallRuleConfig): The parent to be set.
+            """
+            self._parent = parent
+            for rule in self._rules:
+                rule._parent = parent
+
+        def _field_updated(self, attrname):
```

Comment:
Instead of piping a new function throughout the entire SDK why not override the `__set__` function and embedd the code here then call the super class to maintain the same functionality?

Reply:
I actually _considered_ overriding `_set`, the function in `MutableBaseModel` that does the heavy lifting for `__set__`, but there were too many paths through that function that resulted in the field _not_ being actually updated, and I wanted a clear signal as to when it _did_ actually get updated, to minimize unnecessary round-trips to the server when `save` is called on something that hasn't actually _changed._  The best way to do that, in my opinion, was to add the `_field_updated` "hook" to `_set`. If you've got a better way in which this could be implemented by _just_ overriding `_set` without having the hook, I'm all ears.

Reply:
That's why I recommended overriding `__set__` instead of `_set` then calling the super function to maintain existing functionality you may need to override `__setattr__` as well

Reply:
What does overriding `__set__` (and maybe `__setattr__` as well) get me that just overriding `_set` couldn't, in this case?

Looking at the implementation in `MutableBaseModel`, `__setattr__` has several different code paths it can go through. It can go through the property setter objects (which ultimately hit `_set` themselves), it can go through `_set` directly, or it can go through `object.__setattr__`. We don't worry about that last case because it won't be changing anything that is going to get pushed to the server.  Neither it nor `NewBaseModel` actually override `__set__` themselves, so maybe that can be discounted.

Reply:
I guess I am confused why your initial override attempt for `_set` didn't work

Reply:
Hmm...I wonder if an override for `_set` would look like this...
```
def _set(self, attrname, new_value):
    pre_change_value = self._info.get(attrname)
    super(self, classname)._set(attrname, new_value)
    if self._info.get(attrname) != pre_change_value:
        self._mark_changed()
```

Maybe I should try that? It seems a little clunky though.

Reply:
Why do you need check if the change actually took place is there an error condition which would prevent the change?

I would expect the following
```
def _set(self, attrname, new_value):
    super(self, classname)._set(attrname, new_value)
    self._mark_changed()
```

Reply:
But then `_mark_changed` would get called even if the value of that field _didn't actually change._ This defeats the whole purpose of doing this, as we would then be making unnecessary round-trips to the server for fields that didn't actually change.

Reply:
Why is `_set` getting called on properties which are not being changed?

Reply:
Because `_set` gets called for _every_ attempt to set a property value. Inside _that_ method (on or about line 840 in `base.py`) is where it checks for an attempt to set the field to its current value, and it bails out if that's the case.  Therefore, we don't know, _a priori,_ whether the value of the property is going to change inside `_set`, unless we check for it ourselves.

Reply:
I've modified those objects to override `_set` instead and removed the "hook" function altogether.  The test I use actually makes use of the `_dirty_attributes` dict, which the base class `_set` manages. This reduces the required operations to simple `in`, `not in`, and Boolean tests, and reduces the intermediate storage required to a single `bool` value. I think now I'm at the "as simple as possible, but no simpler" stage. This will be committed today.